Enhancement: introduce zip parser to add documents from within zip files (#5570)

bytefactory73 opened the pull request:
implements feature: https://github.com/paperless-ngx/paperless-ngx/discussions/5570

* added special zip handling in 3 places ("# special zip handling")
* security: added special handling of unzipping to not allow install outside the start path (zip slipping)
* tested with zip with sub folders
* tested workflows (are triggered after pdf document added)

<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

<!--
Please include a summary of the change and which issue is fixed (if any) and any relevant motivation / context. List any dependencies that are required for this change. If appropriate, please include an explanation of how your proposed change can be tested. Screenshots and / or videos can also be helpful if appropriate.
-->

<!--
⚠️ Important: Pull requests that implement a new feature or enhancement *should almost always target an existing feature request* with evidence of community interest and discussion. This is in order to balance the work of implementing and maintaining new features / enhancements. If that is not currently the case, please open a feature request instead of this PR to gather feedback from both users and the project maintainers.
-->

Closes #(issue or discussion)

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

bytefactory73 committed:
introduce zip parser to add documents from within zip files (#5570)

https://github.com/paperless-ngx/paperless-ngx/discussions/5570
* added a mime zip type
* added a zip parser that extracts the zip and recursively adds the files from the zip
* added new exception type to have a clean way to skip the zip file which is not really a document but keep it in the history.
* security: added special handling of unzipping to not allow install outside the start path (zip slipping)
* tested with zip with sub folders
* tested workflows (are triggered after pdf document added)

bytefactory73 committed:
fix zip file unittest

bytefactory73 commented:
— 97.47% (-1.60%) compared to 5f8d5a8

what can i do in this case?

bytefactory73 committed:
add documentation

shamoon reviewed (changes requested):
Thanks. Need to think about your approach here, a couple things Im not sure of with the SkipImportException and frankly Im not sure if a parser kicking of `consume_file` tasks is considered breaking the contract or a parser, that would be a good question for our other maintainer

As for testing, you will need to add tests that actually cover your code. You can change the invalid file one or just remove it but just like all parsers there should be tests under paperless_zip/tests/. Im not sure why the coverage wasn't calculated, but for example youre only testing the positive case at the moment, so clearly the PR isnt actually fully covered.

bytefactory73 commented on src/documents/skip_import.py:
any recommendations where to add this?

bytefactory73 commented:
> Thanks. Need to think about your approach here, a couple things Im not sure of with the SkipImportException and frankly Im not sure if a parser kicking of `consume_file` tasks is considered breaking the contract or a parser, that would be a good question for our other maintainer
> 
> As for testing, you will need to add tests that actually cover your code. You can change the invalid file one or just remove it but just like all parsers there should be tests under paperless_zip/tests/. Im not sure why the coverage wasn't calculated, but for example youre only testing the positive case at the moment, so clearly the PR isnt actually fully covered.

i'm not familar with the design of the software. this was the one approach which i thought that it could work because it uses the "plugin/registration" system without modifying internal code which could break other stuff.

another approach could be to implement this directly in consumer.py but this feels a bit more hardcoded.

maybe the core team can suggest the best way here. ..

bytefactory73 committed:
code review fix - added logging for unexpected cases

bytefactory73 commented:
> > i'm not familar with the design of the software. this was the one approach which i thought that it could work because it uses the "plugin/registration" system without modifying internal code which could break other stuff.
> 
> Seriously, no disrespect here but I dont find this very compelling, you are asking to have your code added to the project, taken another way this statement could be read as "I didn't take the time to understand the project". I realize paperless is complex, its a big barrier to contributing, so of course we're not expecting you to be an expert but I'd say in general if you want to open a PR it should respect the existing architecture as much as possible. Again, I'm not trying to scold you or something, just my take on OSS contributing. Regardless, we definitely appreciate the effort.
> 
> > another approach could be to implement this directly in consumer.py but this feels a bit more hardcoded.
> 
> So yea, I guess this is the point, the currently implementation feels a bit hack-ish to me. As you are aware, you create a parser but actually need logic to be run at the consumer level and the parser doesn't do any actual parsing. Again, it would be good to get input from our main backend dev (who unfortunately has been pretty limited on time lately so not sure when he will have the time), but I think handling zips at the consumer level might be better, just kind of pre-processing a zip since after all the actual handling of the files inside will still be up to the actual parsers.

I think you got this in a wrong way. I'm not against changing the implementation. Feel free to tell me how it would be better. I will try to modify the changes ...

shamoon commented:
> I think you got this in a wrong way. I'm not against changing the implementation. Feel free to tell me how it would be better. I will try to modify the changes ...

Yea I didn't mean it harshly, just letting you know how it could be interpreted. Again, we appreciate your willingness to contribute. I think we're in agreement this should be restructured to move zip handling to the consumer level, not a 'fake' parser like the current implementation. Feel free to ask more questions of course but wondering if that's enough of a direction for now? I've marked it as a draft in the meantime, no rush.

bytefactory73 committed:
code review fix - moved the zip extraction code from the zip parser to the consume logic

bytefactory73 commented:
i rewrote the code to move the logic out of the parser into the consume.py

i couldn't find a better way yet. my goal was to:

- let the zip processed as a async job.
- let the added documents be added as well as async jobs.
- the plugin system also operates only on a single per file level from what i found.
- the consume.py is the common central for processing files from where ever they came from

please let me know what you think about this change. is it getting better or worse?

bytefactory73 commented:
> 3. Testing, of course. And as mentioned, you should alter the old .zip test to still do the same thing it currently does (otherwise you are removing a test that actually does test something)
In my opinion we should add a different unsupported file (e.g. test.foo) to have another failing test again

pdf cannot be used anymore since it‘s now a supported type.

bytefactory73 commented:
> 1. Get rid of the zip parser completely, it's just a vestige of the old implementation and makes the whole thing much harder to understand. There is definitely another way to get .zip files into the pipeline (across all sources)

i think removing the pdf parser would mean more hardcoded hacks in several places.

maybe i can find a mechanism how to register the pdf mime type without having a parser registered… hints are welcome

bytefactory73 committed:
Merge branch 'dev' into feature-5570-zip-file-support

bytefactory73 committed:
beautify code

bytefactory73 committed:
reduce code changes to dev branch

bytefactory73 committed:
remove empty files

bytefactory73 committed:
reintroduce the test_upload_invalid_file test

bytefactory73 commented:
> > pdf cannot be used anymore since it‘s now a supported type.
> 
> Indeed, so the test should still test an _unsupported_ type.
> 
> > i think removing the pdf parser would mean more hardcoded hacks in several places.
> > maybe i can find a mechanism how to register the pdf mime type without having a parser registered… hints are welcome
> 
> I suspect thats not the case. A lot of this logic comes from `get_parser_class_for_mime_type` so the question is whether to modify (I think not) that or to look at each place its used and determine which uses actually need to accept a .zip Again, I'm not certain, I'd file that under the work of creating the PR. Happy to continue discussing more before you implement this part if you want to avoid unnecessary work.

@shamoon , @stumpylog 

* i readded the unknown test.
* i reverted the parser code and removed the parser but i introduced special code in 3 places.
* everything works as before

my feeling about this is:
* it's now more hardcoded and hacky. 
* shouldn't we introduce another type of hander, e.g. adding a get_archive_handler() function into the zip_consumer_declaration which can be overloaded in the same way as the parsers do? so we have an non-hacky way of extending the system with arbitrary archive types. so someone else could add tar, gz whatever if needed.

bytefactory73 committed:
fix unittest error

bytefactory73 committed:
fix unittest error

bytefactory73 committed:
fix unittest error

shamoon commented:
> * it's now more hardcoded and hacky.
> * shouldn't we introduce another type of hander, e.g. adding a get_archive_handler() function into the zip_consumer_declaration which can be overloaded in the same way as the parsers do? so we have an non-hacky way of extending the system with arbitrary archive types. so someone else could add tar, gz whatever if needed.

I don't really know what "hard-coded" means in this context especially when comparing to a literal dummy parser. Yes I suppose you could do something like an `archive_handler` but it's hard for me to see a big need for that. On the other hand, with this version:

1. Single source of truth re zip handling with everything in consumer.py
2. No abusing interfaces with dummy parser or pretending zips are actual documents
3. Smaller diff d/t less boilerplate

Personally I think this is better from an architectural standpoint, but I generally defer to stumpylog on the backend, so perhaps he has thoughts. I'll hold off on more specific code review (nothing crazy I think but do have some thoughts) until then (no rush, stumpy).

shamoon commented on docs/faq.md:
```diff
@@ -62,6 +62,8 @@ elsewhere. Here are a couple notes about that.
 -   With the optional Tika integration enabled (see [Tika configuration](https://docs.paperless-ngx.com/configuration#tika)),
     Paperless also supports various Office documents (.docx, .doc, odt,
     .ppt, .pptx, .odp, .xls, .xlsx, .ods).
+-   Zip files are now supported as well. All file types mentioned above
```
```suggestion
-   Zip files are supported as well. All file types mentioned above
```

shamoon commented on docs/faq.md:
```diff
@@ -62,6 +62,8 @@ elsewhere. Here are a couple notes about that.
     Paperless also supports various Office documents (.docx, .doc, odt,
     .ppt, .pptx, .odp, .xls, .xlsx, .ods).
+-   Zip files are now supported as well. All file types mentioned above
+    found in the zip will be extracted and added to paperless.
```
```suggestion
    found in the zip will be extracted and queued for consumption.
```

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
 
             self.log.debug(f"Detected mime type: {mime_type}")
 
+            # special zip handling
```
```suggestion
```

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                        self.input_doc.original_file.unlink()
+                    if self.working_copy.exists():
+                        self.working_copy.unlink()
+                    if (
+                        self.unmodified_original is not None
+                        and self.unmodified_original.exists()
+                    ):  # pragma: no cover
+                        self.unmodified_original.unlink()
```
```suggestion
```

This will never happen for a .zip

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                        consume_file.delay(input_doc, overrides)
+                if tempdir:
+                    tempdir.cleanup()
+                # Explicitly delete the original file and all working copies, but only if they exist
```
```suggestion
```

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                        self.unmodified_original.unlink()
+                except Exception as e:
+                    self.log.warning(f"Could not delete skipped file: {e}")
+                # Mark progress as finished for dashboard/task system
```
```suggestion
```

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                    ):  # pragma: no cover
+                        self.unmodified_original.unlink()
+                except Exception as e:
+                    self.log.warning(f"Could not delete skipped file: {e}")
```
```suggestion
                    self.log.warning(f"Error deleting file: {e}")
```

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                    ProgressStatusOptions.SUCCESS,
+                    ConsumerStatusShortMessage.FINISHED,
+                )
+                return f"Success. File {self.filename} was processed."
```
```suggestion
                return f"Success. Archive {self.filename} was processed."
```

shamoon commented on src/documents/parsers.py:
```diff
@@ -60,6 +60,11 @@ def is_mime_type_supported(mime_type: str) -> bool:
     Returns True if the mime type is supported, False otherwise
     """
+
+    # special zip handling
```
```suggestion
```

shamoon commented on src/documents/parsers.py:
```diff
@@ -108,6 +113,9 @@ def get_supported_file_extensions() -> set[str]:
             # This makes image/webp supported on Python < 3.11
             extensions.add(supported_mime_types[mime_type])
 
+    # special zip handling
```
```suggestion
```

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                        from documents.tasks import consume_file
+
+                        consume_file.delay(input_doc, overrides)
+                if tempdir:
+                    tempdir.cleanup()
```
```suggestion
```

I think better to use a context manager here, e.g. `with tempfile.TemporaryDirectory() as tempdir:` which will automatically cleanup

shamoon commented on src/documents/consumer.py:
```diff
@@ -858,3 +919,44 @@ def run(self) -> None:
+
+def extract_zip_to_tempdir(zip_path, temp_extract_dir, log):
+    """
+    Extracts supported files from a ZIP to temp_extract_dir, returns list of files in the zip.
```
```suggestion
    Extracts files from a ZIP to temp_extract_dir, returns list of files in the zip. Unsupported types
    will be filtered out later at the consumption stage.
```

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                    )
+                    if parser_class is not None:
+                        # Rename file to zipname#originalname before moving
+                        new_name = f"{zip_name}#{file.name}"
```
`file.name` might need sanitizing still?

shamoon commented on src/documents/consumer.py:
```diff
@@ -304,6 +307,64 @@ def run(self) -> str:
+                        # Import here to avoid circular import
+                        from documents.tasks import consume_file
+
+                        consume_file.delay(input_doc, overrides)
```
I think would be better to use `group`, see e.g. 

https://github.com/paperless-ngx/paperless-ngx/blob/5410074062bef9aaaa4979cde7c6a87b39c0d9a3/src/documents/bulk_edit.py#L464
